### PR TITLE
Add support for an lsp language client to specify which capabilities they support.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageClient/XamlInProcLanguageClient.cs [hunk 1]
```diff
@@ -23,8 +23,14 @@ internal class XamlInProcLanguageClient : AbstractInProcLanguageClient
...
         [Obsolete(MefConstruction.ImportingConstructorMessage, true)]
-        public XamlInProcLanguageClient(XamlLanguageServerProtocol languageServerProtocol, VisualStudioWorkspace workspace, IDiagnosticService diagnosticService, IAsynchronousOperationListenerProvider listenerProvider, XamlSolutionProvider solutionProvider)
-            : base(languageServerProtocol, workspace, diagnosticService, listenerProvider, solutionProvider, diagnosticsClientName: null)
+        public XamlInProcLanguageClient(
+            XamlLanguageServerProtocol languageServerProtocol,
+            VisualStudioWorkspace workspace,
+            IDiagnosticService diagnosticService,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            XamlSolutionProvider solutionProvider,
+            DefaultCapabilitiesProvider defaultCapabilitiesProvider)
```
Comment (dibarbet) [inline src/VisualStudio/Xaml/Impl/Implementation/LanguageClient/XamlInProcLanguageClient.cs:32]: the xaml provider needs to specify a different set of capbilities - see their handler here 
http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices.Xaml/Implementation/LanguageServer/Handler/Initialize/InitializeHandler.cs,9663db7ec51dd43e

**Review by dibarbet — COMMENTED** [review]
other than the xaml initialize, everything else looks good

**Review by allisonchou — APPROVED** [review]

**Review by dibarbet — APPROVED** [review]
lgtm, just need to update the tests (maybe we don't even need that test)

**Comment by CyrusNajmabadi** [comment]
I can try that!

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs [hunk 2]
```diff
@@ -91,24 +95,18 @@ internal class InProcLanguageServer
...
-        /// and responding with the server capabilities.
-        /// The specification assures that the initialize request is sent only once.
+        /// Handle the LSP initialize request by storing the client capabilities and responding with the server
+        /// capabilities.  The specification assures that the initialize request is sent only once.
         /// </summary>
         [JsonRpcMethod(Methods.InitializeName, UseSingleObjectParameterDeserialization = true)]
-        public async Task<InitializeResult> InitializeAsync(InitializeParams initializeParams, CancellationToken cancellationToken)
+        public Task<InitializeResult> InitializeAsync(InitializeParams initializeParams, CancellationToken cancellationToken)
         {
+            Contract.ThrowIfTrue(_clientCapabilities != null, $"{nameof(InitializeAsync)} called multiple times");
```
Comment (davidwengier) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:104]: Fun fact: Our code does not live up to this contract check :D

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:104]: wait... how?

Comment (davidwengier) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:104]: We initialize _clientCapabilities in the constructor 🤦‍♂️

I'm fixing it :)

Comment (davidwengier) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:104]: Fixed in https://github.com/dotnet/roslyn/pull/48747

**Pull request merged**